app/forms/schools/add_participant_form.rb

```diff
@@ -6,20 +6,114 @@ class AddParticipantForm
     include ActiveModel::Attributes
     include ActiveModel::Serialization
 
+    STEPS = %i[type details choose_mentor email_taken confirm not_implemented].freeze
+
     TYPE_OPTIONS = {
       ect: "Early Career Teacher",
       mentor: "Mentor",
     }.freeze
 
+    attribute :completed_steps, default: []
     attribute :type
+    attribute :full_name
+    attribute :email
+    attribute :school_cohort_id
+    attribute :mentor_id
 
     validates :type,
               on: :type,
               presence: { message: "Please select type of the new participant" },
-              inclusion: { in: TYPE_OPTIONS.keys.map(&:to_s), allow_blank: true }
+              inclusion: { in: TYPE_OPTIONS.keys, allow_blank: true }
+
+    validates :full_name,
+              on: :details,
+              presence: true
+
+    validates :email,
+              on: :details,
+              presence: true,
+              notify_email: { allow_blank: true }
+
+    validates :mentor_id,
+              on: :choose_mentor,
+              presence: true,
+              inclusion: { in: ->(form) { form.mentor_options.map(&:id) + %w[later] } }
 
     def type_options
       TYPE_OPTIONS
     end
+
+    def mentor_options
+      @mentor_options ||= school_cohort.school.mentors
+    end
+
+    def mentor
+      return @mentor if defined? @mentor
+
+      @mentor = (User.find(mentor_id) if mentor_id.present? && mentor_id != "later")
+    end
+
+    def previous_step(current_step)
+      step_index = completed_steps.index(current_step)
+      return completed_steps.last if step_index.nil?
+      return if step_index.zero?
+
+      completed_steps[step_index - 1]
+    end
+
+    def next_step(step)
+      case step
+      when :type then :details
+      when :details
+        return :email_taken if email_already_taken?
+
+        if type == :ect && mentor_options.any?
+          :choose_mentor
+        else
+          :confirm
+        end
+      when :choose_mentor then :confirm
+      else :not_implemented
+      end
+    end
+
+    def record_completed_step(step)
+      if completed_steps.include? step
+        self.completed_steps = completed_steps[0..completed_steps.index(step)]
+      else
```

Comment:
Yes, but only on the completion of the previous step. Step completion can change the further step order, whcih would seriously iompact the `previous_step` logic. 

imagine user going through type (etc) -> details -> confirmation (completed steps are then `%i[type details]`, then navigating back to `type` and changing selection to `mentor`. In the full journey, it will then be presented with another step to choose whether they want to add an existing mentor from other cohort or the completely new mentor. If they select to create a new mentor they are taken to the details page again. If we didn't trim the `completed steps` on update of `type`, `completed steps` would have value of `%i[type details mentor_options]`, so the back_button on the `confirm` page would lead him to the `mentor_options` instead of `details`.

Ideally, we also should wipe the data provided in the steps we're trimming out